Add flag to temporarily enable first class modules

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21515 Add WeakIValue, use in tracer.
* #21514 clean up the TracingState API
* #21513 Collapse tracing_state.h into tracer.h
* #21325 Interpreter support for CallFunction/CallMethod
* #21273 Expose ExecutionPlan in prep for function calls
* **#21272 Add flag to temporarily enable first class modules**
* #21240 Reduce number of stack manipulation instructions in interpreter.
* #21185 Prepare interpreter for function calling

Summary: this flag is used in tests to enable first-class module
execution so that we can test things like invoking methods on
first-class modules and setting their properties before we implement
all the parts necessary to remove the old execution pattern.

Differential Revision: [D15600068](https://our.internmc.facebook.com/intern/diff/D15600068)

zdevito committed:
Add flag to temporarily enable first class modules

Summary: this flag is used in tests to enable first-class module
execution so that we can test things like invoking methods on
first-class modules and setting their properties before we implement
all the parts necessary to remove the old execution pattern.

zdevito committed:
Update on "Add flag to temporarily enable first class modules"

Add flag to temporarily enable first class modules

Summary: this flag is used in tests to enable first-class module
execution so that we can test things like invoking methods on
first-class modules and setting their properties before we implement
all the parts necessary to remove the old execution pattern.

gh-metadata: pytorch pytorch 21272 gh/zdevito/46/head

zdevito committed:
Update on "Add flag to temporarily enable first class modules"

Add flag to temporarily enable first class modules

Summary: this flag is used in tests to enable first-class module
execution so that we can test things like invoking methods on
first-class modules and setting their properties before we implement
all the parts necessary to remove the old execution pattern.

gh-metadata: pytorch pytorch 21272 gh/zdevito/46/head

Krovatkin commented on test/test_jit.py:
```diff
@@ -3373,6 +3380,21 @@ def __init__(self):
+            foo = Foo()
+            input = torch.rand(3, 4)
+            foo.forward(input)
+            self.assertEqual(input, foo.foo)
```
nitpick: `foo.bar` might be a little bit more readable

Krovatkin commented on torch/csrc/jit/script/module.cpp:
```diff
@@ -213,10 +223,40 @@ static FunctionSchema sliceFirst(const FunctionSchema& schema) {
+
+IValue Method::operator()(std::vector<IValue> stack, const Kwargs& kwargs) {
+  getSchema().checkAndNormalizeInputs(stack, kwargs);
+  if (experimental_run_as_first_class) {
```
nitpick: this logic (pushing module and parameters onto a stack) seems to be duplicated across `run` and `()`. Maybe we should consider refactoring it into a helper, otherwise we might forget to update one of the places eventually?

Krovatkin commented on torch/csrc/jit/script/module.h:
```diff
@@ -57,34 +57,19 @@ using ModuleLookup =
 
 struct TORCH_API Method {
-  Method(Module* owner, Function* function);
+  Method(Module* owner, const std::shared_ptr<Function>& function);
```
so someone else now owns a function aside from `Method`?

zdevito replied:
for now yes, Michael is working on our final state for Function ownership which will make this stuff more clear.

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "Add flag to temporarily enable first class modules"

Add flag to temporarily enable first class modules

Summary: this flag is used in tests to enable first-class module
execution so that we can test things like invoking methods on
first-class modules and setting their properties before we implement
all the parts necessary to remove the old execution pattern.

gh-metadata: pytorch pytorch 21272 gh/zdevito/46/head

Commit:
Update on "Add flag to temporarily enable first class modules"

Add flag to temporarily enable first class modules

Summary: this flag is used in tests to enable first-class module
execution so that we can test things like invoking methods on
first-class modules and setting their properties before we implement
all the parts necessary to remove the old execution pattern.

gh-metadata: pytorch pytorch 21272 gh/zdevito/46/head

zdevito committed:
Update on "Add flag to temporarily enable first class modules"

Add flag to temporarily enable first class modules

Summary: this flag is used in tests to enable first-class module
execution so that we can test things like invoking methods on
first-class modules and setting their properties before we implement
all the parts necessary to remove the old execution pattern.

gh-metadata: pytorch pytorch 21272 gh/zdevito/46/head
